### PR TITLE
Drops negative bucket bounds and negative values in record API

### PR DESCRIPTION
**Pull request by isaikevych**
Issue https://github.com/census-instrumentation/opencensus-node/issues/172

### CODE REVIEW

File: packages/opencensus-core/src/stats/view.ts [hunk 1]
```diff
@@ -59,6 +62,8 @@ export class BaseView implements View {
   endTime: number;
   /** true if the view was registered */
   registered = false;
+  /** An object to log information to */
+  private logger: loggerTypes.Logger = logger.logger();
```
Comment (OsvaldoRosado) [inline packages/opencensus-core/src/stats/view.ts]: Should this be hardcoded? I've noticed in other classes in this repo the pattern seems to be to have a default logger but allow passing one in as an override.

File: packages/opencensus-core/src/stats/view.ts [hunk 2]
```diff
@@ -181,15 +193,55 @@ export class BaseView implements View {
...
-    return bucketBoundaries.map((boundary, boundaryIndex) => {
-      return {
-        count: 0,
-        lowBoundary: boundaryIndex ? boundary : -Infinity,
-        highBoundary: (boundaryIndex === bucketBoundaries.length - 1) ?
-            Infinity :
-            bucketBoundaries[boundaryIndex + 1]
-      };
-    });
+    const boundaries = bucketBoundaries.filter(Math.clz32);
```
Comment (OsvaldoRosado) [inline packages/opencensus-core/src/stats/view.ts]: This doesn't seem to catch all negative numbers. For example, this passes through `-Infinity`. Is that ok?

Comment (OsvaldoRosado) [reply packages/opencensus-core/src/stats/view.ts]: It also seems a little compute wasteful to do this `.filter` just for a console log. Can't the logging done during the reduce?

File: packages/opencensus-core/src/stats/view.ts [hunk 3]
```diff
@@ -181,15 +193,55 @@ export class BaseView implements View {
...
+        this.logger.error('Bucket boundaries not sorted.');
+      }
+      if (current === next) {
+        this.logger.error('Bucket boundaries not unique.');
+      }
+    }
+  }
+
+  /**
+   * Creates empty buckeet boundary.
```
Comment (OsvaldoRosado) [inline packages/opencensus-core/src/stats/view.ts]: Nit: `buckeet` -> `bucket`

File: packages/opencensus-core/src/stats/view.ts (hunk 1 above)
Comment (isaikevych) [reply packages/opencensus-core/src/stats/view.ts]: Fixed for current class, added [issue 177](https://github.com/census-instrumentation/opencensus-node/issues/177) for improve other parts

File: packages/opencensus-core/src/stats/view.ts (hunk 2 above)
Comment (isaikevych) [reply packages/opencensus-core/src/stats/view.ts]: Improved. Thank you

File: packages/opencensus-core/src/stats/view.ts (hunk 3 above)
Comment (isaikevych) [reply packages/opencensus-core/src/stats/view.ts]: Changed

File: packages/opencensus-core/src/stats/view.ts [hunk 4]
```diff
@@ -104,6 +112,12 @@ export class BaseView implements View {
       return;
     }
 
+    if (measurement.value < 0) {
+      this.logger.warn(
+          'Dropping values, value to record must be non-negative.');
```
Comment (songy23) [inline packages/opencensus-core/src/stats/view.ts]: Maybe also log the value?

File: packages/opencensus-core/src/stats/view.ts [hunk 5]
```diff
@@ -181,15 +196,57 @@ export class BaseView implements View {
...
+        const bucket = this.createBucket(boundary, nextBoundary, position, len);
+        accumulator.push(bucket);
+      } else {
+        negative++;
+      }
+      return accumulator;
+    }, []);
+    if (negative) {
+      this.logger.warn(`Dropping ${
+          negative} negative bucket boundaries, the values must be strictly > 0.`);
```
Comment (songy23) [inline packages/opencensus-core/src/stats/view.ts]: ditto

Comment (isaikevych) [reply packages/opencensus-core/src/stats/view.ts]: > ditto

This is what implemented in java [BucketBoundaries.java line 84](https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/stats/BucketBoundaries.java#L84)

Comment (songy23) [reply packages/opencensus-core/src/stats/view.ts]: Ignore this comment - just noticed "${negative}" is already in the log.

File: packages/opencensus-core/src/stats/view.ts (hunk 4 above)
Comment (isaikevych) [reply packages/opencensus-core/src/stats/view.ts]: Done

**Pull request merged**